### PR TITLE
[XPU] Optimize paged FMHA prefill for head dim 512

### PR DESCRIPTION
**Pull request by yuankuns**
## Summary

- port the paged FMHA score-reuse optimization for `HEAD_DIM=512` from CUTLASS-SYCL
- store scaled QK scores in a block2D workspace and reuse them for the second output tile
- preserve remainder-K masking for non-aligned KV lengths, including local-window attention
- keep the existing bf16-only paged prefill dispatch unchanged
- add `FMHA_PREFILL_ENABLE_SCORE_BLOCK2D_512`, enabled by default
- cap the score workspace at 1 GiB by slicing the batch and reusing the same workspace

## Correctness

Tested bf16 on Intel Arc Pro B60 with reverse page tables.

`B=1, SQ=127, SK=250, Hq/Hkv=4/2, D=512`:

| mask | max error | bad elements |
| --- | ---: | ---: |
| full | 0.001964 | 0 |
| causal | 0.002082 | 0 |
| local `(63,17)` | 0.003994 | 0 |

Also tested `B=2` with KV lengths `[250, 193]`, reverse page tables, and local `(63,17)` attention: max error `0.003724`, bad elements `0`.

Additional integration validation:

- full default-feature build: `1485/1485`
- `tests/test_flash_attention.py`: `1725 passed, 660 skipped`
- K-tail, causal GQA, and score-prefetch boundary cases: max absolute error `0.000174-0.000364`, no elements above `0.01`

### Original kernel case

Intel Arc Pro B60, bf16, page size 64, reverse page table, `B=1, SQ=512, SK=4096, Hq/Hkv=16/8, D=512`:

| workload | baseline | optimized | improvement |
| --- | ---: | ---: | ---: |
| full attention | 2.754 ms | 2.203 ms | 20.0% |
| Gemma4 global attention (causal) | 2.615 ms | 2.138 ms | 18.2% |

### Large production shape

Intel Arc Pro B60, bf16, page size 128, `B=16, SQ=SK=4096, Hq/Hkv=16/16, D=512`. The baseline is the saved pre-PR wheel; both wheels used the same benchmark script on GPU 1 with `triton.testing.do_bench`, `warmup=1000 ms`, and `rep=5000 ms`.

| workload | baseline | optimized | latency improvement | baseline TOPS | optimized TOPS |
| --- | ---: | ---: | ---: | ---: | ---: |
| full attention | 297.427 ms | 175.081 ms | 41.1% (`1.70x`) | 29.57 | 50.24 |
| causal attention | 153.899 ms | 95.652 ms | 37.8% (`1.61x`) | 28.58 effective | 45.99 effective |

For this shape, batch slicing reduces the score workspace from 8 GiB to 1 GiB (`batch_slice=2`, 8 sequential slices).

### Gemma4 31B TP=4 E2E

Gemma4 31B, Intel XPU attention, GPUs 4/5/6/7, `TP=4`, random input/output lengths `4096/1024`, concurrency 1, one warmup request, radix cache disabled, chunked prefill disabled, and profiler disabled. The baseline and optimized runs used the same SGLang worktree and server/client arguments; only the `sgl_kernel` wheel changed.

| metric | baseline wheel | optimized wheel | change |
| --- | ---: | ---: | ---: |
| TTFT | 1522.11 ms | 1490.95 ms | -2.05% |
| E2E latency | 109.46 s | 107.50 s | -1.79% |
| TPOT | 105.51 ms | 103.63 ms | -1.79% |
| mean ITL | 105.61 ms | 103.73 ms | -1.78% |
| output throughput | 9.35 token/s | 9.52 token/s | +1.83% |
| total throughput | 46.76 token/s | 47.61 token/s | +1.83% |

### CODE REVIEW

**Review by airMeng — APPROVED** [review]
Let's wait for model level perf results

**Comment by Valentine233** [comment]
@yuankuns @airMeng @sunjiweiswift This PR causes a hang issue with reproducer:
```
diff --git a/tests/test_flash_attention.py b/tests/test_flash_attention.py
index 490c9b2..0cce2da 100644
--- a/tests/test_flash_attention.py
+++ b/tests/test_flash_attention.py
@@ -480,9 +480,11 @@ def generate_qkv(
     [torch.bfloat16, torch.float16]
     + ([torch.float8_e4m3fn] if not DISABLE_FP8 else []),
 )
-@pytest.mark.parametrize("nheads_q,nheads_kv", [(16, 16), (16, 4)])
+# @pytest.mark.parametrize("nheads_q,nheads_kv", [(16, 16), (16, 4)])
+@pytest.mark.parametrize("nheads_q,nheads_kv", [(8, 1)])
 @pytest.mark.parametrize("new_kv", [False])
-@pytest.mark.parametrize("causal,local", [(False, True), (False, False), (True, False)])
+# @pytest.mark.parametrize("causal,local", [(False, True), (False, False), (True, False)])
+@pytest.mark.parametrize("causal,local", [(True, False)])
 @pytest.mark.parametrize("use_sinks", [True, False])
 @pytest.mark.parametrize("seqlen_new_eq_seqlen_q", [True])
 @pytest.mark.parametrize("has_rotary_seqlens", [False])
@@ -497,22 +499,30 @@ def generate_qkv(
         else [0.0]
     ),
 )
-@pytest.mark.parametrize("page_size", [64, 128])
+# @pytest.mark.parametrize("page_size", [64, 128])
+@pytest.mark.parametrize("page_size", [128])
 @pytest.mark.parametrize("has_leftpad", [False])
 @pytest.mark.parametrize("has_batch_idx", [False])
 @pytest.mark.parametrize("varlen_q", [True])
-@pytest.mark.parametrize("d", [64, 128, 256, 512])
+# @pytest.mark.parametrize("d", [64, 128, 256, 512])
+@pytest.mark.parametrize("d", [512])
+# @pytest.mark.parametrize(
+#     "seqlen_q,seqlen_k",
+#     [
+#         (3, 1024),
+#         (64, 800),
+#         (64, 256),
+#         (3, 799),
+#         (64, 2048),
+#         (128, 128),
+#         (256, 512),  # To test appending KV with more than 1 block
+#         (2048, 3577),  # Enough tile to test persistent scheduler
+#     ],
+# )
 @pytest.mark.parametrize(
     "seqlen_q,seqlen_k",
     [
-        (3, 1024),
-        (64, 800),
-        (64, 256),
-        (3, 799),
-        (64, 2048),
-        (128, 128),
-        (256, 512),  # To test appending KV with more than 1 block
-        (2048, 3577),  # Enough tile to test persistent scheduler
+        (512, 512),
     ],
 )
 def test_flash_attn_kvcache(
```
Building with `-DFMHA_PREFILL_ENABLE_SCORE_BLOCK2D_512=OFF`, the test could pass.

**Comment by yuankuns** [comment]
@Valentine233 Can you share the oneAPI/umd/kmd version for the reproducer?

**Comment by yuankuns** [comment]
@Valentine233 @airMeng @sunjiweiswift Just confirmed that b3c1a497a13984f9ba3c939f6a0a2bf5e6210d76 (commit before pr342) also trigger hang when num_head_q/num_head_kv=8/16, these ratioes are never covered by ut.

**Comment by yuankuns** [comment]
Just run a cross matrix scan on b3c1a49 (main before pr342)
image
it will hang at page=128 and headdim=512

File: src/sycl/kernels/flash_attention_v2/kernel/xe_fmha_fwd_kernel.hpp [hunk 1]
```diff
@@ -179,22 +194,62 @@ class XeFMHAFwdKernel {
...
+        score_k_extent = size_t(args.kernel.shape.seq_len_kv_cache.max_length);
+      } else {
+        score_q_extent = size_t(args.kernel.shape.seq_len_qo);
+        score_k_extent = size_t(args.kernel.shape.seq_len_kv_cache);
+      }
+      const size_t score_rows_per_wg = size_t(get<0>(TileShapeQK{}));
+      const size_t q_tiles = cute::ceil_div(score_q_extent, score_rows_per_wg);
+      score_k_extent = cute::round_up(score_k_extent, size_t(get<1>(TileShapeQK{})));
+      return size_t(args.kernel.shape.batch) * size_t(args.kernel.shape.num_heads_q) * q_tiles * score_rows_per_wg *
+             score_k_extent * sizeof(typename CollectiveMainloop::ElementScoreStore);
```
Comment (Valentine233) [inline src/sycl/kernels/flash_attention_v2/kernel/xe_fmha_fwd_kernel.hpp:251]: When the input shape is large, OOM issue happens here.
Could we limit the shapes for `ScoreBlock2D`?
Reproduced config:
```
@pytest.mark.parametrize("nheads_q,nheads_kv", [(16, 2)])
@pytest.mark.parametrize("has_qv", [False])
@pytest.mark.parametrize("deterministic", [False])
@pytest.mark.parametrize("softcap", [0.0] + ([15.0] if not DISABLE_SOFTCAP else []))
@pytest.mark.parametrize("causal,local", [(True, False)])
@pytest.mark.parametrize("add_unused_qkv", [False])
@pytest.mark.parametrize("d", [512])
@pytest.mark.parametrize(
    "seqlen_q,seqlen_k",
    [
        (32768, 32768),
    ],
)
def test_flash_attn_varlen_output(
    batch_size = 1
```
@yuankuns Please help have a look!
